### ISSUE TITLE
[P0-BUG] optimizer.step() hangs at reduce_scatter_tensor — systematic NCCL collective mismatch in DistributedOptimizer

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#154]
## Symptoms
Training hangs reproducibly at `optimizer.step()` after forward+backward complete:
```
rank=0: ENTERING optimizer.step
(hangs forever, Ctrl+C needed)
```

## Root Cause Analysis

`DistributedOptimizer.step()` calls `prepare_grads()` → `_reduce_scatter_grads()` which issues `torch.distributed.reduce_scatter_tensor()` on `self.data_parallel_group`. This is a **collective** that requires ALL ranks in the DP group to call it simultaneously.

### Why it deadlocks

The training loop in `desloc_engine.py` wraps `optimizer.step()` inside `if not _should_skip:`. If one rank decides `_should_skip=True` (via HeteroGradNormSkipController) while another decides `_should_skip=False`, the stepping rank enters `reduce_scatter_tensor` but the skipping rank never arrives → permanent NCCL hang.

Additional collective mismatch sources:
1. `finalize_model_grads()` calls `_allreduce_conditional_embedding_grads()` + `_allreduce_router_expert_bias_grads()` — collectives that ALL ranks must enter
2. `clip_grad_norm()` calls `torch.distributed.all_reduce()` for grad norm computation
3. `shard_to_model_broadcast()` after optimizer.step — implicit allgather

### Prior fix attempts
- `0dfd3628` — removed `num_microbatches` allreduce, added barrier
- `d8251080` — switched to `dist_optimizer.step()` + disabled `force_all_reduce`
- `6dabfbe9` — added `cuda.synchronize` before `AdamW.step()`

None addressed the fundamental issue: **the skip decision must be collective**.

## Proposed Fix

1. All-reduce `_should_skip` flag BEFORE branching on it — ensure all ranks agree
2. Even when skipping the Adam update, ALL ranks must still enter reduce_scatter_grads (with zero contribution) to avoid NCCL mismatch
3. Add `NCCL_TIMEOUT` config so hangs fail fast instead of hanging forever

## Acceptance
- Training completes step 0-10 without hang on 3-GPU setup (2×A6000 + 1×H100)
- No NCCL timeout errors
- Gradient skip controller still functional

**Comment by dylanyunlon** [comment]
Fixed in 79938243.

**Root cause**: `_should_skip` was computed locally per rank. On heterogeneous hardware, numerical differences cause rank divergence → one rank enters `reduce_scatter_tensor` while another skips → NCCL deadlock.

**Fix (two parts)**:
1. All-reduce `_should_skip` (ReduceOp.MAX) before branching — if ANY rank skips, ALL skip
2. Decomposed `dist_optimizer.step()` into `prepare_grads()` (always entered) + `step_with_ready_grads()` (conditional) — both NCCL collectives always entered by all ranks

This supersedes prior fix attempts: 0dfd3628, d8251080, 6dabfbe9.
